src/cwgen/cwgen.1
.\"
.\" UnixCW CW Tutor Package - CWGEN
.\" Copyright (C) 2001-2006  Simon Baldwin (simon_baldwin@yahoo.com)
.\" Copyright (C) 2011-2015  Kamil Ignacak (acerion@wp.pl)
.\"
.\" This program is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License
.\" as published by the Free Software Foundation; either version 2
.\" of the License, or (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\"
.TH CWGEN 1 "CW Tutor Package" "cwgen ver. 3.4.2" \" -*- nroff -*-
.SH NAME
.\"
cwgen \- generate groups of random characters for Morse code practice
.\"
.\"
.\"
.SH SYNOPSIS
.\"
.B cwgen
[\-g\ \-\-groups=\fIgroups\fP]
[\-n\ \-\-groupsize=\fIgroupsize\fP]
[\-n\ \-\-groupsize=\fIgroup_min-group_max\fP]
[\-r\ \-\-repeat=\fIrepeat\fP]
[\-x\ \-\-limit=\fIlimit\fP]
[\-c\ \-\-charset=\fIcharset\fP]
.BR
[\-h\ \-\-help]
[\-V\ \-\-version]
.PP
\fBcwgen\fP installed on GNU/Linux systems understands both short form
and long form command line options.  \fBcwgen\fP installed on other
operating systems may understand only the short form options.
.PP
Options may be predefined in the environment variable \fBCWGEN_OPTIONS\fP.
If defined, these options are used first; command line options take
precedence.
.PP
.\"
.\"
.\"
.SH DESCRIPTION
.\"
.PP
.B cwgen
provides a simple method to generate groups of random characters for
use as input into \fBcw\fP.  Command line options control the group
size, number of groups, and character set from which the random
characters are selected.
.PP
.\"
.\"
.\"
.SS COMMAND LINE OPTIONS
.\"
.B cwgen
understands the following command line options.  The long form options
may not be available in non-LINUX versions.
.TP
.I "\-g, \-\-groups"
Specifies the number of groups of random characters to generate.  The
default value is 128.
.TP
.I "\-n, \-\-groupsize"
Specifies either the number of random characters in each group, if a
single value is given, or a range for the number of random characters
in each group if a dash-separated pair of values is given.  If a range
is given, \fBcwgen\fP generates groups sized randomly between 'group_min'
and 'group_max'.  The default value is 5.
.TP
.I "\-r, \-\-repeat"
Specifies the number of times to repeat each group.  The default value
is 0, indicating that each group is printed just once.
.TP
.I "\-x, \-\-limit"
Specifies an upper limit on the number of characters printed.  This option
is useful when using random group sizes, to ensure a given amount of
practice text.  The default value is 0, indicating no upper limit on the
number of characters printed.
.TP
.I "\-c, \-\-charset"
Defines the character set from which the random characters are
selected.  The default value is 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'.
.PP
.\"
.\"
.\"
.SH EXAMPLES
.\"
Generate 20 groups of 10 characters from the set EISH5, and sound
them as Morse code.  Both short-form and long-form option examples
are shown:
.IP
cwgen \-g 20 \-n 10 \-c "EISH5" | cw \-w 25 \-t 850
.IP
cwgen \-\-groups=20 \-\-groupsize=10 \-\-charset="EISH5" |
cw \-\-wpm=25 \-\-tone=850
.PP
.\"
.\"
.\"
.SH SEE ALSO
.\"
Man pages for \fBcw\fP(7,LOCAL), \fBlibcw\fP(3,LOCAL), \fBcw\fP(1,LOCAL),
\fBcwcp\fP(1,LOCAL), and \fBxcwcp\fP(1,LOCAL).
.\"


NEWS
version 3.4.2 / 2015-06-17

  libcw:

  - fixing one possible cause of segfaults in generator code
    (segfaults occurred when pthread_kill() was called with invalid
    thread ID in some specific circumstances).

  - fixing another problem with libcw: when a generator is created,
    then is *not* started, and then is deleted, libcw called
    pthread_kill(&id, ...) on uninitialized id in
    cw_gen_delete(). This caused a program to crash.

  - fixing problems with OSS and generator modules that surfaced
    during tests on Alpine Linux.

  - expanding unit tests.

  - libcw is now compiled with -DNDEBUG flag by default (asserts are
    "off"). The asserts are still "on" by default in "unit tests"
    libcw. Asserts are switched "on" in base libcw with "./configure
    --enable-dev".

  - cw_generator_set_tone_slope(): some corner cases of handling
    function's arguments were cleared in function's top level
    comments, and implementation has been updated accordingly.

  - Library soname/version changed from 6.3.1 to 6.4.1.


----------------------------------------------------------------------------


version 3.4.1 / 2015-01-02

  libcw:

  - The remainder of code from libcw.c has been moved to respective
    module files. New module file has been created: libcw_rec.c - it
    contains receiver code. I think that number of modules won't
    change anymore. libcw.c has been removed, all functionality is now
    implemented by module files.

  - A number of unit test functions has been moved from
    libcw_test_public.c to respective module files. These functions
    can be executed independently, so there is no reason to keep them
    in libcw_test_public.c. They can be kept in module files, close to
    tested library functions. libcw_test_public.c is now less
    difficult to work with. The moved test functions are now executed
    while building "make check" target, so they should be called more
    often, increasing the chances of catching problems early.

  - Data sets for some of receiver's unit tests are now generated by
    helper functions. This allowed me to increase size of the data
    sets and increase variation of the data (both for fixed speed
    receiving and for adaptive receiving).

    The data sets are basically timing tables used to drive the
    receiver - to test main functions of the receiver, responsible for
    accepting mark/space events from client code.

    The increased size and variation of the test data gives me more
    confidence that the core functionality of the receiver works as
    expected.

  - If you run libcw in debugging mode and you depend on Receiver
    State names (strings starting with RS_) printed by library's debug
    messages, you may want to check changed values of RS enum and
    corresponding strings on top of libcw_rec.c - they have been
    changed.

  - Some changes have been made to library functions (mainly in
    generator and receiver modules) to ensure that the code separated
    between modules can be compiled and that it works as expected.
    Since receiver code has been put in a separate file, it was
    possible to review and refactor the receiver code a bit.

    No functionality of the library has been changed, with an
    exception described in next point.

  - Data type of receiver's speed is now internally a float instead of
    int. This change improves behaviour of adaptive receiving
    algorithm a bit. It's not a big change, but it may positively
    impact client code using receiver functions.

  - libcw_test_public test executable now accepts '-m' command line
    option (for selecting module(s)) to test. Check output of
    'libcw_test_public -h' for more details.

  - Since the modification described above does not change how the
    library works, nor does it change any API, there is no reason for
    any major revision changes. Library soname/version has been
    changed from 6.2.1 to 6.3.1.


----------------------------------------------------------------------------


version 3.4.0 / 2014-11-11

  unixcw:

  - Build system files have been modified to correctly configure and
    compile unixcw (or at least parts of it) on following platforms
    (in addition to Debian GNU/Linux):
    - FreeBSD 10.0 (x86_64-unknown-freebsd10.0, with clang 3.3);
    - FreeBSD 9.3 (freebsd9.3 with gcc 4.2.1);
    - OpenBSD 5.5 (amd64-unknown-openbsd5.5 with gcc 4.2.1);

    See INSTALL file for more information.

  - QT4 configuration flags (CFLAGS, LIBS, MOC) are now detected only
    by using pkg-config. My hand-made script that searched for QT4
    files has been reduced to simple call to PKG_CHECK_MODULES().
    This means that the following arguments are no longer accepted by
    configure script:
    --with-qt-cflags=<string>
    --with-qt-libraries=<path>

    If you want to override pkg-config settings, you can still use
    environment variables. Check output of "./configure --help" for
    details


  libcw:

  - Large parts of libcw.c have been moved to new files:
    - libcw_gen.c (generator)
    - libcw_tq.c (tone queue)
    - libcw_data.c (representations and lookup)
    - libcw_key.c (straight key, iambic keyer)
    - libcw_signal.c (signal handling)
    - libcw_utils.c (utility functions)

    Benefits of this move:
    - I won't have to use a table of contents to move around in
      libcw.c (the file had 10kLOC before this operation);
    - it will be easier to define boundaries between functionality of
      modules, and it will be easier to reason about how these modules
      should function and interact with each other;

    Check ChangeLog entry from 2014-07-28 for more information on this
    subject.

    Separation of libcw modules is not yet completed. Receiver code is
    still mixed with generator code, and there are still some
    functions left in libcw.c. I expect to finish the separation in
    one or two development cycles.


src/cwgen/cwgen.c
/*
 * Copyright (C) 2001-2006  Simon Baldwin (simon_baldwin@yahoo.com)
 * Copyright (C) 2011-2015  Kamil Ignacak (acerion@wp.pl)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h> /* gettimeofday() */
#include <errno.h>
#include <assert.h>
#include <stdint.h>
#include <inttypes.h> /* SCNu64 in sscanf() */

#if defined(HAVE_STRING_H)
# include <string.h>
#endif

#if defined(HAVE_STRINGS_H)
# include <strings.h>
#endif

#include "i18n.h"
#include "cmdline.h"
#include "cw_copyright.h"
#include "memory.h"





#define MIN_GROUPS             1   /* Lowest number of groups allowed. */
#define INITIAL_GROUPS       128   /* Default number of groups. */
#define MIN_GROUP_SIZE         1   /* Lowest group size allowed. */
#define INITIAL_GROUP_SIZE     5   /* Default group size. */
#define INITIAL_REPEAT         0   /* Default repeat count. */
#define MIN_REPEAT             0   /* Lowest repeat count allowed. */
#define MIN_LIMIT              0   /* Lowest character count limit allowed. */
#define INITIAL_LIMIT          0   /* Default character count limit. */


static const char *const DEFAULT_CHARSET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";


struct cwgen_config {
	char *program_name;    /* Program's name (argv[0]) */

	int n_groups;          /* How many unique groups to generate. */
	int group_size_min;    /* Minimal size (length) of generated groups. */
	int group_size_max;    /* Maximal size (length) of generated groups. */
        int n_repeats;         /* How many times in a row to repeat each unique group; zero for no repeats (group will be printed only once). */
	uint64_t n_chars_max;  /* Maximal number of characters (excluding spaces) to generate in whole set of groups; may be zero - no limit. */

	char *charset;         /* Set of chars to be used to generate groups. */
} config = {
	.program_name   = (char *) NULL,

	.n_groups       = INITIAL_GROUPS,
	.group_size_min = INITIAL_GROUP_SIZE,
        .group_size_max = INITIAL_GROUP_SIZE,
        .n_repeats      = INITIAL_REPEAT,
        .n_chars_max    = INITIAL_LIMIT,

	.charset        = (char *) NULL
};


static const char *all_options = "g:|groups,n:|groupsize,r:|repeat,x:|limit,c:|charset,h|help,v|version";

static void cwgen_generate_characters(struct cwgen_config *config);
static void cwgen_print_usage(const char *program_name);
static void cwgen_print_help(const char *program_name);
static void cwgen_parse_command_line(int argc, char **argv, struct cwgen_config *config);
static void cwgen_free_config(struct cwgen_config *config);




/**
   \brief Generate random characters on stdout

   Generate random characters on stdout, in groups as requested, and up
   to the requested number of groups.  Characters are selected from the
   set given at random.

   \param config - program's configuration variable
*/
void cwgen_generate_characters(struct cwgen_config *config)
{
	static bool is_initialized = false;

	/* On first (usually only) call, seed the random number generator. */
	if (!is_initialized) {

		/* Previously the initializator used value returned by
		time(). Consecutive calls of the program within one
		second resulted in the same generated string - not
		very random. To improve randomness, I've switched to
		gettimeofday(). */
		struct timeval t;
		gettimeofday(&t, NULL);

		srandom(t.tv_usec);
		is_initialized = true;
	}

	/* Allocate the buffer for repeating groups. */
	char *buffer = (char *) malloc(config->group_size_max);
	if (!buffer) {
		fprintf(stderr, "%s: failed to allocate memory\n", config->program_name);
		exit(EXIT_FAILURE);
	}

	/* Generate groups up to the number requested or to the character limit. */
	int charset_length = strlen(config->charset);
	uint64_t chars = 0;

	for (int group = 0; group < config->n_groups; group++) {

		/* Randomize the group size between min and max inclusive. */
		int group_size = config->group_size_min + random() % (config->group_size_max - config->group_size_min + 1);

		/* Create random group. */
		for (int i = 0; i < group_size; i++) {
			buffer[i] = config->charset[random() % charset_length];
		}

		/* Repeatedly print the group as requested.
		   It's always printed at least once, then repeated
		   for the desired repeat count.  Break altogether if
		   we hit any set limit on printed characters. */
		int repeat = 0;
		do {
			for (int i = 0; i < group_size; i++) {

				putchar(buffer[i]);
				fflush(stdout);

				chars++;

				if (config->n_chars_max && chars >= config->n_chars_max) {
					break;
				}
			}

			putchar(' ');
			fflush(stdout);

			if (config->n_chars_max && chars >= config->n_chars_max) {
				break;
			}

		} while (repeat++ < config->n_repeats);

		if (config->n_chars_max && chars >= config->n_chars_max) {
			break;
		}
	}

	free(buffer);

	return;
}





/**
   \brief Print out a brief message directing the user to the help function

   \param program_name - program's name
*/
void cwgen_print_usage(const char *program_name)
{
	const char *format = has_longopts()
		? _("Try '%s --help' for more information.\n")
		: _("Try '%s -h' for more information.\n");

	fprintf(stderr, format, program_name);
	return;
}





/*
  \brief Print out a brief page of help information

  \param program_name - program's name
*/
static void cwgen_print_help(const char *program_name)
{
	if (!has_longopts()) {
		fprintf(stderr, "%s", _("Long format of options is not supported on your system\n\n"));
	}

	printf(_("Usage: %s [options...]\n\n"), program_name);

	printf(_("  -g, --groups=GROUPS    send GROUPS groups of chars [default %d]\n"), INITIAL_GROUPS);
	printf(_("                         GROUPS values may not be lower than %d\n"), MIN_GROUPS);
	printf(_("  -n, --groupsize=GS     make groups GS chars [default %d]\n"), INITIAL_GROUP_SIZE);
	printf(_("                         GS values may not be lower than %d, or\n"), MIN_GROUP_SIZE);
	printf("%s", _("  -n, --groupsize=GL-GH  make groups between GL and GH chars\n"));
	printf("%s", _("                         valid GL, GH values are as for GS above\n"));
	printf(_("  -r, --repeat=COUNT     repeat each group COUNT times [default %d]\n"), INITIAL_REPEAT);
	printf(_("                         COUNT values may not be lower than %d\n"), MIN_REPEAT);
	printf("%s", _("  -c, --charset=CHARSET  select chars to send from this set\n"));
	printf(_("                         [default %s]\n"), DEFAULT_CHARSET);
	printf(_("  -x, --limit=LIMIT      stop after LIMIT characters [default %d]\n"), INITIAL_LIMIT);
	printf("%s", _("                         a LIMIT of zero indicates no set limit\n"));
	printf("%s", _("  -h, --help             print this message\n"));
	printf("%s", _("  -v, --version          output version information and exit\n\n"));

	exit(EXIT_SUCCESS);
}





/**
   \brief Parse command line options

   Parse the command line options for initial values for the various
   global and flag definitions.

   \param argc - main()'s argc
   \param argv - main()'s argv
   \param config - program's configuration variable
*/
void cwgen_parse_command_line(int argc, char **argv, struct cwgen_config *config)
{
	int option;
	char *argument;

	config->program_name = strdup(cw_program_basename(argv[0]));
	if (!config->program_name) {
		fprintf(stderr, "%s: failed to allocate memory\n", argv[0]);
		exit(EXIT_FAILURE);
	}

	while (get_option(argc, argv, all_options,
			  &option, &argument)) {

		switch (option) {
		case 'g':
			if (sscanf(argument, "%d", &(config->n_groups)) != 1
			    || config->n_groups < MIN_GROUPS) {

				fprintf(stderr, _("%s: invalid groups value: '%s'\n"), config->program_name, argument);
				exit(EXIT_FAILURE);
			}
			break;

		case 'n':
			if (sscanf(argument, "%d-%d", &(config->group_size_min), &(config->group_size_max)) == 2) {
				if (config->group_size_min < MIN_GROUP_SIZE
				    || config->group_size_max < MIN_GROUP_SIZE
				    || config->group_size_min > config->group_size_max) {

					fprintf(stderr, _("%s: invalid groupsize range: '%s'\n"), config->program_name, argument);
					exit(EXIT_FAILURE);
				}
			} else if (sscanf(argument, "%d", &(config->group_size_min)) == 1) {
				if (config->group_size_min < MIN_GROUP_SIZE) {

					fprintf(stderr, _("%s: invalid groupsize value\n"), config->program_name);
					exit(EXIT_FAILURE);
				}
				config->group_size_max = config->group_size_min;
			}
			break;

		case 'r':
			if (sscanf(argument, "%d", &(config->n_repeats)) != 1
			    || config->n_repeats < MIN_REPEAT) {

				fprintf(stderr, _("%s: invalid repeat value: '%s'\n"), config->program_name, argument);
				exit(EXIT_FAILURE);
			}
			break;

		case 'x':
			if (sscanf(argument, "%" SCNu64, &(config->n_chars_max)) != 1
			    || strstr(argument, "-")) {

				fprintf(stderr, _("%s: invalid limit value: %s\n"), config->program_name, argument);
				exit(EXIT_FAILURE);
			}
			fprintf(stderr, _("%s: valid limit value: %s\n"), config->program_name, argument);
			break;

		case 'c':
			if (strlen(argument) == 0) {
				fprintf(stderr, _("%s: charset cannot be empty\n"), config->program_name);
				exit(EXIT_FAILURE);
			}
			assert(!config->charset);
			config->charset = strdup(argument);
			if (!config->charset) {
				fprintf(stderr, _("%s: failed to allocate memory\n"), config->program_name);
				exit(EXIT_FAILURE);
			}
			break;

		case 'h':
			cwgen_print_help(config->program_name);

		case 'v':
			printf(_("%s version %s\n%s\n"),
			       config->program_name, PACKAGE_VERSION, _(CW_COPYRIGHT));
			exit(EXIT_SUCCESS);

		case '?':
			cwgen_print_usage(config->program_name);
			exit(EXIT_FAILURE);

		default:
			fprintf(stderr, _("%s: getopts returned %c\n"), config->program_name, option);
			exit(EXIT_FAILURE);
		}
	}

	if (get_optind() != argc) {
		cwgen_print_usage(config->program_name);
		exit(EXIT_FAILURE);
	}

	return;
}





/**
   \brief Parse the command line options, then generate the characters requested
*/
int main(int argc, char **argv)
{
	int combined_argc;
	char **combined_argv;

	/* Set locale and message catalogs. */
	i18n_initialize();

	/* Parse combined environment and command line arguments. */
	combine_arguments(_("CWGEN_OPTIONS"),
			  argc, argv, &combined_argc, &combined_argv);
	cwgen_parse_command_line(combined_argc, combined_argv, &config);

	if (!config.charset) {
		config.charset = strdup(DEFAULT_CHARSET);
		if (!config.charset) {
			fprintf(stderr, _("%s: failed to allocate memory\n"), argv[0]);
			return EXIT_FAILURE;
		}
	}

	/* Generate the character groups as requested. */
	cwgen_generate_characters(&config);
	putchar('\n');

	cwgen_free_config(&config);

	return EXIT_SUCCESS;
}





/**
   \brief Deallocate memory used by fields of config variable

   Function calls free() for all pointers to previously allocated memory.

   \param config - pointer to config variable
*/
void cwgen_free_config(struct cwgen_config *config)
{
	if (config->charset) {
		free(config->charset);
		config->charset = (char *) NULL;
	}

	if (config->program_name) {
		free(config->program_name);
		config->program_name = (char *) NULL;
	}

	return;
}
